internal/core/processor/universal/processor.go

```diff
@@ -81,24 +87,128 @@ func (p *processor) VerifyChainConsistency(ctx context.Context, startingBlock ui
 		slog.Error("error occurred while reading LaosEvolution end range block hash", "err", err.Error())
 		return err
 	}
-
 	// During the initial iteration, no hash is stored in the database, so this code block is bypassed.
 	if (lastBlockDB.Hash == common.Hash{}) {
 		return nil
 	}
+	return p.checkBlockForReorg(ctx, lastBlockDB)
+}
+
+func (p *processor) RecoverFromReorg(ctx context.Context, currentBlock uint64) (*model.Block, error) {
+	// Start a transaction
+	tx := p.stateService.NewTransaction()
+	defer tx.Discard()
+
+	storedBlockNumbers, err := tx.GetAllStoredBlockNumbers()
```

Comment:
I think from this line  `storedBlockNumbers, err := tx.GetAllStoredBlockNumbers()` until line 118 code should be placed in one function. For instance and that function can be named `getValidBlock` block for which reorg did not happen. 

I think this code

`nextBlockNumberToCheck, err := getNextLowerBlockNumber(currentBlock, storedBlockNumbers)
	if err != nil { // no lower block number found
		// we get a safe block number to start from
		blockWithoutReorg = getSafeBlock(currentBlock)
`

    should go inside p.CheckForReorgRecursive. I see that function also calls `getSafeBlock` .

Not mandatory from my side but I recomend to give it a try.